src/librustc_mir/interpret/memory.rs

```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     pub tcx: TyCtxtAt<'a, 'tcx, 'tcx>,
 }
 
+impl<'a, 'mir, 'tcx, M> Clone for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn clone(&self) -> Self {
+        Memory {
+            data: self.data.clone(),
+            alloc_kind: self.alloc_kind.clone(),
+            alloc_map: self.alloc_map.clone(),
+            uninitialized_statics: self.uninitialized_statics.clone(),
+            cur_frame: self.cur_frame.clone(),
+            tcx: self.tcx.clone(),
+        }
+    }
+}
+
+impl<'a, 'mir, 'tcx, M> Eq for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{}
+
+impl<'a, 'mir, 'tcx, M> PartialEq for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn eq(&self, other: &Self) -> bool {
+        let Memory {
+            data,
+            alloc_kind,
+            alloc_map,
+            uninitialized_statics,
+            cur_frame,
+            tcx,
+        } = self;
+
+        *data == other.data
+            && *alloc_kind == other.alloc_kind
+            && *alloc_map == other.alloc_map
+            && *uninitialized_statics == other.uninitialized_statics
+            && *cur_frame == other.cur_frame
+            && ptr::eq(tcx, &other.tcx)
+    }
+}
+
+impl<'a, 'mir, 'tcx, M> Hash for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let Memory {
+            data,
+            alloc_kind: _,
+            alloc_map: _,
+            uninitialized_statics: _,
+            cur_frame,
+            tcx,
+        } = self;
+
+        data.hash(state);
+        cur_frame.hash(state);
+        (tcx as *const _ as usize).hash(state);
+
+        // We ignore some fields which don't change between evaluation steps.
```

Comment:
You aren't really ignoring them, hashing the `allocations` iterator should get you most of the info.

Reply:
This ignores `uninitialized_statics` at the moment, which actually can change during execution, but I think will not affect infinite loop detection, since modifying an uninitialized static would cause it to no longer be initialized. I could be way off here though.

Reply:
That's dangerous. Equality and hashing need to be in sync.  I don't remember the exact rules, but you can read up on that in the hash docs

Reply:
`uninitialized_statics` (along with another field) is ignored in the hash, but not in the equality check. This maintains the invariant required by `Hash`, that two objects with different hashes never compare as equal.

This should be fixed anyways, but I need to know how to compare `FxHashMap<AllocId, Allocation>`s against one another.

Reply:
Right. That makes sense